### PR TITLE
feat(core): in memory mempool

### PR DESCRIPTION
**Pull request by jrchatruc**
**Motivation**

This PR aims to move the mempool (i.e. transaction pool) from being a `libmdbx` table to an in-memory data structure. 
This makes sense as the mempool is not something meant to be persisted for a long time, but rather a temporary place for transactions for (ideally) a few seconds/minutes until they're included in blocks or discarded. If the node goes down and then back up for whatever reason, mempool transactions will come back from the p2p gossip network anyway.

More concretely, though, the mempool being on disk was being a huge bottleneck on the load tests we were running, which were sending approximately 5k transactions per second to the node. When doing this with the current on-disk mempool, the node would hang and completely stop executing new blocks until the load test finished sending all transactions, at which point it would resume execution. With this (very naive and rushed) in-memory implementation, the node does not hang and can keep executing blocks at a normal pace while also receiving and storing all incoming transactions.

Things to fix in the future:

- Limiting the size of the mempool. We'd like to avoid the possibility of an OOM due to too many transactions (note that this was also a problem with the on-disk structure, running out of disk space instead of memory though).
- Optimizing the way transactions are stored, filtered and picked to be included on each block. To build a block, the node currently does the following:

  - Keep a data structure of all the mempool transactions, grouped by sender address and sorted by nonce: for each sender address, we want transactions ordered by nonce in increasing order to know what the next transaction is for an account quickly. This is the `txs` field of the `TransactionQueue` struct right now.
  - Keep a data structure of all the `head` transactions sorted by fee tip in decreasing order, i.e., a list that has, for each sender address, its first transaction in terms of nonce, with the highest paying one first and so on (this is just a subset of the above where we just keep the first transaction for each key on a list and then sort that by fee). This is the `heads` field of the `TransactionQueue`.
  - While there's still gas left in the block, take the first transaction from the head transactions list and execute it; for that transaction's sender, take their next transaction (in terms of nonce) and add it to the head transactions list, making sure it remains sorted after insertion. Repeat until the block is full.

  This ensures the node builds a block with all valid transactions (i.e. no invalid nonces) and with the highest possible fee for them. However right now there's a lot of copying going on because the mempool is kept separately from the `TransactionQueue`; to build the `TransactionQueue` the builder first takes each mempool tx (stored on the `mempool` field of the `Store`) and passes it through a filter, then copies the remains of that filter to the Queue and builds the sorted head transactions list from there.

  My impression is that a lot of this copying and duplicating of transactions can be removed.

### CODE REVIEW

File: crates/storage/store/storage.rs [hunk 1]
```diff
@@ -210,51 +218,87 @@ impl Store {
...
+
+        match maybe_old_value {
+            Some(old_value) => {
+                old_value.insert(transaction.nonce(), transaction);
+            }
+            None => {
+                let mut new_value = BTreeMap::new();
+                new_value.insert(transaction.nonce(), transaction);
+                mempool.insert(sender, new_value);
+            }
```
Comment (fmoletta) [inline crates/storage/store/storage.rs:234]: Could we replace this match with `entry(..).or_default().insert(..)`?

File: crates/storage/store/storage.rs [hunk 2]
```diff
@@ -210,51 +218,87 @@ impl Store {
...
+            .get(&tx_hash)
+            .cloned()
     }
 
     /// Remove a transaction from the pool
-    pub fn remove_transaction_from_pool(&self, hash: H256) -> Result<(), StoreError> {
-        self.engine.remove_transaction_from_pool(hash)
+    pub fn remove_transaction_from_pool(&self, address: Address, nonce: u64) {
+        if let Some(old_value) = self.mempool.lock().unwrap().get_mut(&address) {
+            let tx = old_value.get(&nonce).unwrap();
```
Comment (fmoletta) [inline crates/storage/store/storage.rs:258]: Do we know for sure that the transaction exists here? Maybe we can obtain it from the `remove` method

File: crates/blockchain/mempool.rs [hunk 3]
```diff
@@ -100,8 +94,8 @@ pub fn filter_transactions(
 }
 
 /// Remove a transaction from the mempool
-pub fn remove_transaction(hash: H256, store: &Store) -> Result<(), StoreError> {
-    store.remove_transaction_from_pool(hash)
+pub fn remove_transaction(address: Address, nonce: u64, store: &Store) {
```
Comment (mpaulucci) [inline crates/blockchain/mempool.rs:97]: why are you indexing with address + nonce? Technically there could be more than 1 tx with the same address and nonce.

File: crates/storage/store/storage.rs [hunk 4]
```diff
@@ -210,51 +218,87 @@ impl Store {
...
-    /// Get a transaction from the pool
-    pub fn get_transaction_from_pool(
-        &self,
-        hash: H256,
-    ) -> Result<Option<MempoolTransaction>, StoreError> {
-        self.engine.get_transaction_from_pool(hash)
+    pub fn add_transaction_to_pool(&self, transaction: MempoolTransaction) {
+        let mut mempool = self.mempool.lock().unwrap();
+        let maybe_old_value = mempool.get_mut(&transaction.sender());
+        let sender = transaction.sender();
```
Comment (ManuelBilbao) [inline crates/storage/store/storage.rs:230]: We should call `sender()` as few as posible, as it consumes a lot of time (I think it doesn't cache it). Maybe it's better to extract it to a variable

File: crates/blockchain/mempool.rs (hunk 3 above)
Comment (jrchatruc) [reply crates/blockchain/mempool.rs:97]: `address` here means sender address, I'll change that to make it more clear. 

I don't believe there can be more than 1 tx with same sender and nonce, if someone sends a second tx with the same nonce we either reject it if the fee is lower or replace the current one with it if the fee is higher

File: crates/storage/store/storage.rs (hunk 1 above)
Comment (ilitteri) [reply crates/storage/store/storage.rs:240]: Great suggestion!

File: crates/storage/store/storage.rs (hunk 4 above)
Comment (mpaulucci) [reply crates/storage/store/storage.rs]: we can probably just cache it

File: crates/blockchain/mempool.rs [hunk 5]
```diff
@@ -54,31 +54,26 @@ pub fn add_transaction(transaction: Transaction, store: Store) -> Result<H256, M
...
 pub fn get_blobs_bundle(tx_hash: H256, store: Store) -> Result<Option<BlobsBundle>, MempoolError> {
-    Ok(store.get_blobs_bundle_from_pool(tx_hash)?)
+    store
+        .get_blobs_bundle_from_pool(tx_hash)
+        .map_err(MempoolError::from)
 }
 
 /// Applies the filter and returns a set of suitable transactions from the mempool.
-/// These transactions will be grouped by sender and sorted by nonce
+/// These transactions will be grouped by sender and sorted by hash.
```
Comment (mpaulucci) [inline crates/blockchain/mempool.rs:65]: sorted by hash doesn't make much sense IMO.

File: crates/blockchain/mempool.rs [hunk 6]
```diff
@@ -100,8 +96,8 @@ pub fn filter_transactions(
 }
 
 /// Remove a transaction from the mempool
-pub fn remove_transaction(hash: H256, store: &Store) -> Result<(), StoreError> {
-    store.remove_transaction_from_pool(hash)
+pub fn remove_transaction(address: Address, hash: H256, store: &Store) -> Result<(), StoreError> {
+    store.remove_transaction_from_pool(address, hash)
```
Comment (mpaulucci) [inline crates/blockchain/mempool.rs:100]: if you're passing the hash, why do you need the address?

File: crates/storage/store/storage.rs (hunk 2 above)
Comment (ilitteri) [reply crates/storage/store/storage.rs]: Is this what you meant [730aa10](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1002/commits/730aa101663d7d0150a8cd7696ead00f67e905cc)?

File: crates/blockchain/mempool.rs (hunk 3 above)
Comment (ilitteri) [reply crates/blockchain/mempool.rs]: This has been updated, we now index with `sender_address` and `tx_hash`.

File: crates/storage/store/storage.rs (hunk 4 above)
Comment (ilitteri) [reply crates/storage/store/storage.rs]: Refactored this piece of code, now we call it once. This has improved a little bit the flamegraph but it is still a bottle neck.

File: crates/blockchain/mempool.rs (hunk 5 above)
Comment (ilitteri) [reply crates/blockchain/mempool.rs]: This has been updated.

File: crates/blockchain/mempool.rs (hunk 6 above)
Comment (ilitteri) [reply crates/blockchain/mempool.rs]: The mempool is now a `Hash<Address, MempoolTransactionStore>`, to either add or remove a transaction to it we need the sender's address for the first indexing there.

File: crates/storage/store/storage.rs (hunk 2 above)
Comment (fmoletta) [reply crates/storage/store/storage.rs]: Yes!

**Review by fkrause98 — APPROVED** [review]

**Pull request merged**